eng/bomgenerator/src/main/java/com/azure/tools/bomgenerator/DependencyAnalyzer.java

```diff
@@ -0,0 +1,238 @@
+package com.azure.tools.bomgenerator;
+
+import com.azure.tools.bomgenerator.models.BomDependency;
+import com.azure.tools.bomgenerator.models.BomDependencyNoVersion;
+import org.jboss.shrinkwrap.resolver.api.maven.Maven;
+import org.jboss.shrinkwrap.resolver.api.maven.MavenFormatStage;
+import org.jboss.shrinkwrap.resolver.api.maven.MavenResolvedArtifact;
+import org.jboss.shrinkwrap.resolver.api.maven.MavenResolverSystemBase;
+import org.jboss.shrinkwrap.resolver.api.maven.MavenStrategyStage;
+import org.jboss.shrinkwrap.resolver.api.maven.PomEquippedResolveStage;
+import org.jboss.shrinkwrap.resolver.api.maven.PomlessResolveStage;
+import org.jboss.shrinkwrap.resolver.api.maven.ScopeType;
+import org.jboss.shrinkwrap.resolver.api.maven.coordinate.MavenDependency;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+import static com.azure.tools.bomgenerator.Utils.RESOLVED_EXCLUSION_LIST;
+
+public class DependencyAnalyzer {
+    private Set<BomDependency> inputDependencies = new HashSet<>();
+    private Set<BomDependency> externalDependencies = new HashSet<>();
+    private Set<BomDependency> bomEligibleDependencies = new HashSet<>();
+    private Map<BomDependency, String> bomIneligibleDependencies = new HashMap<>();
+    private Map<BomDependencyNoVersion, BomDependency> coreDependencies = new HashMap<>();
+
+    private Map<BomDependencyNoVersion, HashMap<String, Collection<BomDependency>>> nameToVersionToChildrenDependencyTree = new TreeMap<>(new Comparator<BomDependencyNoVersion>() {
+        @Override
+        public int compare(BomDependencyNoVersion o1, BomDependencyNoVersion o2) {
+            return (o1.getGroupId() + o1.getArtifactId()).compareTo(o1.getGroupId() + o2.getArtifactId());
+        }
+    });
+    private static Logger logger = LoggerFactory.getLogger(BomGenerator.class);
+
+    DependencyAnalyzer(Collection<BomDependency> inputDependencies, Collection<BomDependency> externalDependencies){
+        if(inputDependencies != null) {
+            this.inputDependencies.addAll(inputDependencies);
+        }
+        if(externalDependencies != null) {
+            this.externalDependencies.addAll(externalDependencies);
+        }
+    }
+
+    public Collection<BomDependency> getBomEligibleDependencies() {
+        return this.bomEligibleDependencies;
+    }
+
+    public void reduce() {
+        analyze();
+        generateReport();
+        this.bomEligibleDependencies.retainAll(this.inputDependencies);
+    }
+
+    public boolean validate() {
+        analyze();
+        return nameToVersionToChildrenDependencyTree.keySet().stream().anyMatch(key -> nameToVersionToChildrenDependencyTree.get(key).size() > 1);
```

Comment:
This could be simplified to just use `values()` which is a Collection of the values in the map:

```java
return nameToVersionToChildrenDependencyTree.values().stream().anyMatch(value -> value.size() > 1);
```

Reply:
Thx!